Sources/NIOSSL/SSLCertificateExtensions.swift

```diff
@@ -0,0 +1,221 @@
+    private enum Storage {
+        final class Deallocator {
+            let reference: OpaquePointer?
+            
+            init(takeOwnershipOf reference: OpaquePointer?) {
+                self.reference = reference
+            }
+            
+            deinit {
+                if let reference = self.reference {
+                    CNIOBoringSSL_sk_X509_EXTENSION_free(reference)
+                }
+            }
+        }
+        
+        case owned(Deallocator)
+        case borrowed(reference: OpaquePointer?, owner: AnyObject)
+        
+        init(takeOwnershipOf reference: OpaquePointer?) {
+            self = .owned(.init(takeOwnershipOf: reference))
+        }
+        
+        init(borrowing reference: OpaquePointer?, owner: AnyObject) {
+            self = .borrowed(reference: reference, owner: owner)
+        }
+        
+        /// The owner of the memory to which the reference points
+        var owner: AnyObject {
+            switch self {
+            case .owned(let deallocator):
+                return deallocator
+            case .borrowed(_, let owner):
+                return owner
+            }
+        }
+        
+        /// All operations accessing `reference` need to be implemented while guaranteeing that we still have a reference to the memory owner.
+        /// Otherwise `reference` could already be freed. This would result in undefined behaviour as we access a dangling pointer.
+        /// This method guarantees that `reference` is valid during execution of `body`.
+        internal func withReference<Result>(
+            _ body: (OpaquePointer?) throws -> Result
+        ) rethrows -> Result {
+            try withExtendedLifetime(self) {
+                switch self {
+                case .owned(let deallocator):
+                    return try body(deallocator.reference)
+                case .borrowed(let reference, _):
+                    return try body(reference)
+                }
+            }
+        }
+    }
+    
+    @usableFromInline internal let stackSize: Int
+    private let storage: Storage
+    
+    internal init(takeOwnershipOf reference: OpaquePointer?) {
+        self.storage = .init(takeOwnershipOf: reference)
+        if let reference = reference {
+            self.stackSize = CNIOBoringSSL_sk_X509_EXTENSION_num(reference)
+        } else {
+            self.stackSize = 0
+        }
+    }
+    
+    internal init(borrowing reference: OpaquePointer?, owner: AnyObject) {
+        self.storage = .init(borrowing: reference, owner: owner)
+        if let reference = reference {
+            self.stackSize = CNIOBoringSSL_sk_X509_EXTENSION_num(reference)
+        } else {
+            self.stackSize = 0
+        }
+    }
+    
+    internal init(copyOf reference: OpaquePointer?) {
+        guard let reference = reference else {
+            self.init()
+            return
+        }
+        self.init(takeOwnershipOf: CNIOBoringSSL_sk_X509_EXTENSION_dup(reference))
```

Comment:
It is possible for this to return `nil`. I'm inclined to say we should trap in that case: getting a silent copy of nothing seems not great.

Reply:
Actually do we use this at all? If not we should probably just drop it.